src/blockMonitor/blockchainMachine.ts

```diff
@@ -31,62 +106,92 @@ export class BlockchainMachine<TBlock extends IBlockStub> extends StartStopServi
+                    await this.blockItemStore.putBlockItem(block.number, block.hash, `${component.name}:prevEmittedState`, prevHeadAnchorState);
+                }
+            }
+        } finally {
+            this.lock.release();
         }
     }
 
-    private processNewHead(head: Readonly<TBlock>, prevHead: Readonly<TBlock> | null) {
-        // The components can specify some behaviour that is computed as a diff
-        // between the old head and the head. We compute this now for each of the
-        // components
-
-        for (const { component, states } of this.componentsAndStates) {
-            const state = states.get(head);
-            if (state == undefined) {
-                // Since processNewBlock is always called before processNewHead, this should never happen
-                this.logger.error(
-                    `State for component ${component.constructor.name} for block ${head.hash} (number ${
-                        head.number
-                    }) was not set, but it should have been.`
-                );
-                return;
-            }
-            if (prevHead) {
-                const prevState = states.get(prevHead);
-                if (prevState) {
-                    const actions = component.detectChanges(prevState, state);
+    private async processNewHead(head: Readonly<TBlock>) {
+        try {
+            await this.lock.acquire();
+
+            // The components can specify some behaviour that is computed as a diff
+            // between the old head and the head. We compute this now for each of the
+            // components
+
+            for (const component of this.components) {
+                const state: AnchorState = this.blockItemStore.getItem(head.hash, `${component.name}:state`);
+                if (state == undefined) {
+                    // Since processNewBlock is always called before processNewHead, this should never happen
+                    this.logger.error(
+                        `State for component ${component.constructor.name} for block ${head.hash} (number ${head.number}) was not set, but it should have been.`
+                    );
+                    return;
+                }
+
+                const prevEmittedState: AnchorState | null = this.blockItemStore.getItem(head.hash, `${component.name}:prevEmittedState`);
+
+                // this is now the latest anchor stated for an emitted head block; update the store accordingly
+                await this.blockItemStore.putBlockItem(head.number, head.hash, `${component.name}:prevEmittedState`, state);
+
+                if (prevEmittedState) {
+                    // save actions in the store
+                    const newActions = component.detectChanges(prevEmittedState, state);
+                    if (newActions.length > 0) await this.actionStore.storeActions(component.name, newActions);
+
+                    // load all the actions (might include oldler actions; also, they now have id)
```

Comment:
If we're loading older actions isnt it possible that if we have to quick consecutive blocks we fire all the actions for the component twice?

Reply:
As processNewBlock and processNewHead in the BlockchainMachine now acquire a lock, it is guaranteed that events processing never happen in parallel in the BlockchainMachine, so we should be good - unless I missed something.

Reply:
I see, what I was missing is that indeed the actions run in their separate "thread". Good point.
As the issue is quite serious (and perhaps easy to miss), I spent some time to create a failing testcase in 75a3572cbeb2e798f373368eaa91caa4a81f478e, which I then fixed in 53c1919432c2fc28d99381d83194aa5cad7bd9f8 by keeping track of the ids of running actions.

Reply:
Further refactoring done in f99b463, including some minor refactoring and documentation of the ActionStore.